### PR TITLE
Add address, amount and line item fields to Venmo createPaymentContext

### PR DESCRIPTION
**Pull request by khushboo18**
Thank you for your contribution to Braintree. 

> Before submitting this PR, note we cannot accept language translation PRs. We support the same languages that are supported by PayPal, and have a dedicated localization team to provide the translations. If there is an error in a specific translation, you may open an issue and we will escalate it to the localization team.

### Summary of changes

Adds the following fields to BTVenmoRequest to be used in createPaymentContext:

2 new address collection fields -

collectCustomerBillingAddress
collectCustomerShippingAddress

and 6 new fields for transaction amount and line items -

lineItems
subTotalAmount
discountAmount
taxAmount
shippingAmount
totalAmount

Note: Should not be merged until [this GraphQL PR](https://github.braintreeps.com/braintree/cosmos-apps/pull/8614) is merged.

### JIRA
https://paypal.atlassian.net/browse/DTBTVENMO-940

### Checklist

- [x] Added a changelog entry

- khushboo18

### CODE REVIEW

**Review by sshropshire — APPROVED** [review]
🌊

File: Braintree.xcodeproj/project.pbxproj
Comment (jaxdesmarais) [inline Braintree.xcodeproj/project.pbxproj:1418]: We typically sort our files by name. If you right click the directory in Xcode you should see the option "sort by name" towards the bottom

File: CHANGELOG.md
Comment (jaxdesmarais) [inline CHANGELOG.md:4]: General question, why are these fields being added? Is this a feature or just so merchants can send additional details in the request?

Comment (jaxdesmarais) [reply CHANGELOG.md:4]: Typically we'd put this under a section with the module name like:
```
* BraintreeVenmo
    * heading and properties nested here 
```

File: Sources/BraintreeCore/BTConfiguration+Extensions.swift
Comment (jaxdesmarais) [inline Sources/BraintreeCore/BTConfiguration+Extensions.swift:43]: Heads up that this will need to move into the `BTConfiguration+Venmo` file once this PR is merged in: https://github.com/braintree/braintree_ios/pull/1032

File: Sources/BraintreeVenmo/BTVenmoClient.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoClient.swift:113]:
```suggestion
            var paysheetDetails: [String: String] = [
                "collectCustomerBillingAddress": "\(request.collectCustomerBillingAddress)",
                "collectCustomerShippingAddress": "\(request.collectCustomerShippingAddress)"
            ]
```

File: Sources/BraintreeVenmo/BTVenmoClient.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoClient.swift:150]:
```suggestion
                if let jsonAmountData = try? JSONSerialization.data(withJSONObject: transactionDetails),
```

File: Sources/BraintreeVenmo/BTVenmoClient.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoClient.swift:157]:
```suggestion
                if let paysheetDetailsData = try? JSONSerialization.data(withJSONObject: paysheetDetails),
```

File: Sources/BraintreeVenmo/BTVenmoError.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoError.swift:59]: We should add this new error case to the CHANGELOG

File: Sources/BraintreeVenmo/BTVenmoLineItem.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoLineItem.swift:68]: Can this be internal? It looks like it's only used in this module.

File: Sources/BraintreeVenmo/BTVenmoLineItem.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoLineItem.swift:74]:
```suggestion
        ] as [String: Any]
```

**Review by jaxdesmarais — APPROVED** [review]
Curious on what the timeline is on the GraphQL PR being merged in.

**Review by jaxdesmarais — COMMENTED** [review]
Curious on what the timeline is on the GraphQL PR being merged in.

File: Sources/BraintreeCore/BTConfiguration+Extensions.swift
Comment (khushboo18) [reply Sources/BraintreeCore/BTConfiguration+Extensions.swift]: done!

File: CHANGELOG.md
Comment (khushboo18) [reply CHANGELOG.md:4]: Its a feature. Merchants can add pass these fields and based on that, new info (amounts and line items) will be displayed on the Venmo Paysheet, as well as collection of shipping and billing addresses.

I'll update the CHANGELOG file.

File: Sources/BraintreeVenmo/BTVenmoLineItem.swift
Comment (khushboo18) [reply Sources/BraintreeVenmo/BTVenmoLineItem.swift:68]: This is used here in the Venmo Client so I believe it cannot be internal - https://github.com/khushboo18/braintree_ios/blob/cb4058db8e9fad6c2be866088854fdab9dd512a2/Sources/BraintreeVenmo/BTVenmoClient.swift#L142

File: Braintree.xcodeproj/project.pbxproj
Comment (khushboo18) [reply Braintree.xcodeproj/project.pbxproj:1451]: Hmm I can't seem to find this file `project.pbxproj` in my local repository... 🤔

**Comment by khushboo18** [comment]
Thanks for the review @jaxdesmarais! 
The graphQL PR is awaiting another PR to get merged so that we can verify the integration tests. I would say it should be good to merge in about a week or so.

File: Sources/BraintreeVenmo/BTVenmoLineItem.swift
Comment (scannillo) [reply Sources/BraintreeVenmo/BTVenmoLineItem.swift:68]: Since the file it is accessed from also lives in the same module (BraintreeVenmo) we can use internal here! Internal is the default in Swift, so we can just remove the `public` keywork.

File: Braintree.xcodeproj/project.pbxproj
Comment (scannillo) [inline Braintree.xcodeproj/project.pbxproj:10]: Looks like this file was incorrectly added to the Tests section

File: CHANGELOG.md
Comment (scannillo) [inline CHANGELOG.md:14]:
```suggestion
```

This doesn't need a CHANGELOG entry since it is not merchant facing

File: Sources/BraintreeVenmo/BTConfiguration+Venmo.swift
Comment (scannillo) [inline Sources/BraintreeVenmo/BTConfiguration+Venmo.swift:30]: What does "enriched customer data" mean/do?

File: Sources/BraintreeVenmo/BTVenmoLineItem.swift
Comment (scannillo) [inline Sources/BraintreeVenmo/BTVenmoLineItem.swift:7]:
```suggestion
```

File: Sources/BraintreeVenmo/BTVenmoRequest.swift
Comment (scannillo) [inline Sources/BraintreeVenmo/BTVenmoRequest.swift:54]: Nitpicks for consistency

```suggestion
    /// Optional. The subtotal amount of the transaction to be displayed on the paysheet. Excludes taxes, discounts, and shipping amounts.
```

File: Sources/BraintreeVenmo/BTVenmoRequest.swift
Comment (scannillo) [inline Sources/BraintreeVenmo/BTVenmoRequest.swift:62]:
```suggestion
    /// Optional. The total tax amount for the transaction to be displayed on the paysheet.
```

File: Sources/BraintreeVenmo/BTVenmoRequest.swift
Comment (scannillo) [inline Sources/BraintreeVenmo/BTVenmoRequest.swift:70]:
```suggestion
    /// Optional. The grand total amount on the transaction that should be displayed on the paysheet.
```

File: Sources/BraintreeVenmo/BTVenmoRequest.swift
Comment (scannillo) [inline Sources/BraintreeVenmo/BTVenmoRequest.swift:74]:
```suggestion
    /// Optional. The line items for this transaction. It can include up to 249 line items.
```

File: Sources/BraintreeVenmo/BTConfiguration+Venmo.swift
Comment (khushboo18) [reply Sources/BraintreeVenmo/BTConfiguration+Venmo.swift:30]: This is a checkbox that lives in the Control Panel. When checked, it allows merchants to receive user info (name, email, and phone number as of now) during the consent flow.
In the context of this PR, merchants are being allowed to pass 2 flags - `collectCustomerBillingAddress` and `collectCustomerShippingAddress`. These addresses can only be collected when this ECD checkbox is checked in the CP.

File: CHANGELOG.md
Comment (jaxdesmarais) [inline CHANGELOG.md:5]: General thought for merchant/sales/support clarity. We should add a line item between the heading and properties about this feature. Something like `Allow merchants to collect enriched customer data if enabled`. This way if someone needs to know when we added this feature it's easy to find in our changelog.

File: Sources/BraintreeVenmo/BTConfiguration+Venmo.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTConfiguration+Venmo.swift:30]: I think we should opt to use the full name vs acronym in our documentation - that way people don't need to hunt down what `ECD` means

File: Sources/BraintreeVenmo/BTVenmoError.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoError.swift:59]:
```suggestion
    /// Enriched Customer Data is disabled
    case enrichedCustomerDataDisabled
```

File: Sources/BraintreeVenmo/BTVenmoError.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoError.swift:111]:
```suggestion
            return "Cannot collect customer data when Enriched Customer Data is disabled."
```
Can we add something like "Enable this feature in the Control Panel to collect this data." to the end of this message? That assumes it's self service for merchants to enable. If it's not we should add a not about contact support/sales/whomever is responsible for enabling this feature.

File: Sources/BraintreeVenmo/BTVenmoLineItem.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoLineItem.swift:57]:
```suggestion
    // MARK: - Internal Methods
```

File: UnitTests/BraintreeVenmoTests/BTConfiguration+Venmo_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreeVenmoTests/BTConfiguration+Venmo_Tests.swift:44]:
```suggestion
            "payWithVenmo": ["enrichedCustomerDataEnabled": true]
```

File: UnitTests/BraintreeVenmoTests/BTConfiguration+Venmo_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreeVenmoTests/BTConfiguration+Venmo_Tests.swift:51]:
```suggestion
            "payWithVenmo": ["enrichedCustomerDataEnabled": false]
```

File: UnitTests/BraintreeVenmoTests/BTConfiguration+Venmo_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreeVenmoTests/BTConfiguration+Venmo_Tests.swift:59]:
```suggestion
        configurationJSON = BTJSON(value: ["payWithVenmo": [:]])
```

File: UnitTests/BraintreeVenmoTests/BTVenmoClient_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreeVenmoTests/BTVenmoClient_Tests.swift:126]:
```suggestion
           let input = inputDict["input"] as? [String: Any] {
```

File: UnitTests/BraintreeVenmoTests/BTVenmoClient_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreeVenmoTests/BTVenmoClient_Tests.swift:185]:
```suggestion
           let input = inputDict["input"] as? [String: Any] {
```

File: UnitTests/BraintreeVenmoTests/BTVenmoClient_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreeVenmoTests/BTVenmoClient_Tests.swift:212]:
```suggestion
        let input = inputDict["input"] as? [String: Any] {
```

File: Sources/BraintreeVenmo/BTVenmoError.swift
Comment (khushboo18) [reply Sources/BraintreeVenmo/BTVenmoError.swift:111]: Good idea 👍🏼

File: CHANGELOG.md
Comment (jaxdesmarais) [inline CHANGELOG.md:5]:
```suggestion
  * Allow merchants to collect enriched customer data if enabled in the Braintree Control Panel
```

File: Sources/BraintreeVenmo/BTVenmoLineItem.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoLineItem.swift:67]:
```suggestion
        var requestParameters: [String: Any] = [
            "quantity": quantity,
            "unit_amount": unitAmount,
            "name": name,
            "kind": kind == .debit ? "debit" : "credit"
        ]
```

**Comment by scannillo** [comment]
Before we merge this, we'll want to do manual end-to-end test of the flow with these params set (via [this demo view controller](https://github.com/braintree/braintree_ios/blob/master/Demo/Application/Features/Venmo%20-%20Custom%20Button/BraintreeDemoCustomVenmoButtonViewController.m)).

We could also consider updating our demo feature to set all of these params as well.

File: Sources/BraintreeCore/BTGraphQLHTTP.swift
Comment (khushboo18) [inline Sources/BraintreeCore/BTGraphQLHTTP.swift:206]: We want to add specific error messaging for the validation errors coming from the payment context service, and those do not have an `inputPath`

File: Demo/Application/Features/Venmo - Custom Button/BraintreeDemoCustomVenmoButtonViewController.m
Comment (jaxdesmarais) [inline Demo/Application/Features/Venmo - Custom Button/BraintreeDemoCustomVenmoButtonViewController.m:40]:
```suggestion
    BTVenmoLineItem *lineItem = [[BTVenmoLineItem alloc] initWithQuantity:1 unitAmount:@"30" name:@"item-1" kind: BTVenmoLineItemKindDebit];
    [lineItem setUnitTaxAmount:@"1"];
    [venmoRequest setLineItems:[NSArray arrayWithObject: lineItem]];

```

File: Podfile.lock
Comment (jaxdesmarais) [inline Podfile.lock:48]: We will want to revert the changes to `Podfile.lock` in this PR

File: Sources/BraintreeCore/BTGraphQLHTTP.swift
Comment (jaxdesmarais) [inline Sources/BraintreeCore/BTGraphQLHTTP.swift:209]:
```suggestion
                    if let message = errorJSON["message"].asString() {
                        return BTGraphQLErrorTree(message: message).toDictionary()
```

File: Sources/BraintreeVenmo/BTVenmoClient.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoClient.swift:91]:
```suggestion
                notifyFailure(with: BTVenmoError.enrichedCustomerDataDisabled, completion: completion)
```
This will ensure we send the expected analytics events for this client

File: Sources/BraintreeVenmo/BTVenmoClient.swift
Comment (jaxdesmarais) [inline Sources/BraintreeVenmo/BTVenmoClient.swift:159]: Since we are passing through the error now we can do an `if let error` on the line above to avoid the double unwrapping here. If let will check that the value is `!= nil`.

**Review by jaxdesmarais — COMMENTED** [review]
Just a few general cleanup comments. Overall this is looking good! 👏

File: Podfile.lock
Comment (khushboo18) [reply Podfile.lock:48]: Yeah, I wasn't sure how these came up. Will revert.

File: Sources/BraintreeVenmo/BTVenmoClient.swift
Comment (khushboo18) [reply Sources/BraintreeVenmo/BTVenmoClient.swift:159]: Do you mean something like this: 
```
if let error = error as? NSError {
      let jsonResponse: BTJSON? = error.userInfo[BTCoreConstants.jsonResponseBodyKey] as? BTJSON
      ...
}
```
Just to clarify, we are only passing through the error msg only in the case of venmo payment context service validation errors.

**Comment by khushboo18** [comment]
@scannillo @jaxdesmarais can we get another review & approval on this PR please? This has been tested and is ready to merge now. TIA!

File: Braintree.xcodeproj/project.pbxproj
Comment (jaxdesmarais) [inline Braintree.xcodeproj/project.pbxproj:1607]: we can revert this, these files should be sorted by name

File: CHANGELOG.md
Comment (scannillo) [inline CHANGELOG.md:30]: We ended up favoring having all of our error code updates be in this[ SDK_ERROR_CODES.md file](https://github.com/braintree/braintree_ios/blob/master/SDK_ERROR_CODES.md) instead of the CHANGELOG.

I can make that update for you, but don't have push access to your fork.

**Review by scannillo — APPROVED** [review]
Thanks for all of your work in this PR 👏 ! Left a small CHANGELOG comment, otherwise LGTM.

File: Braintree.xcodeproj/project.pbxproj
Comment (sshropshire) [reply Braintree.xcodeproj/project.pbxproj]: We should try [this](https://stackoverflow.com/a/32470246) one day. I wonder how well it works.

File: CHANGELOG.md
Comment (khushboo18) [reply CHANGELOG.md]: Thanks @scannillo. I'm noticing that the Error file does not have any `BTVenmoError`s listed under the `BraintreeVenmo` section. There's one `BTVenmoAppSwitchError` but it's missing some other existing BTVenmoError codes listed [here](https://github.com/khushboo18/braintree_ios/blob/c5500d05b69b0bb7b371161c7f1ff35be1f5c153/Sources/BraintreeVenmo/BTVenmoError.swift#L29). Are those left out on accident?

Comment (jaxdesmarais) [reply CHANGELOG.md]: Nope, we just missed those, not intentionally left off, they should absolutely be on there. 🤦

Comment (khushboo18) [reply CHANGELOG.md]: Sounds good! I copied over those changes to this PR as well in case we want to merge it all together. If not, will rebase once your PR goes through @jaxdesmarais

Comment (jaxdesmarais) [reply CHANGELOG.md]: Merged in the other PR earlier today

Comment (khushboo18) [reply CHANGELOG.md]: Rebased on the latest master :)

File: CHANGELOG.md
Comment (jaxdesmarais) [inline CHANGELOG.md:19]:
```suggestion
```

We can remove this since it's internal

File: CHANGELOG.md
Comment (jaxdesmarais) [inline CHANGELOG.md:8]: Can also move this grouping above the breaking changes section - this is all additions there is nothing that is considered a breaking change in this PR

File: CHANGELOG.md
Comment (khushboo18) [reply CHANGELOG.md:19]: Uhhh I was pretty sure I removed it but all this rebasing and trying to maintain commit history is chaotic 😅

**Comment by jaxdesmarais** [comment]
👋 @khushboo18 -

Wanted to make sure this was highlighted in it's own release, so we will release these changes next sprint vs having it get lost in the v6 GA release. Would you mind updating the changelog one last time to move these changes into an `## unreleased` section above the 6.0.0 release section? Since I don't have push access to your fork I can't make this change.

**Comment by scannillo** [comment]
UI tests were failing in unrelated module (BTAmericanExpress). Those are also known flake failures. Merged into master!

**Comment by scannillo** [comment]
👋 @khushboo18 This was released in[ v6.1.0](https://github.com/braintree/braintree_ios/releases/tag/6.1.0)!

**Pull request merged**
